homeassistant/components/sensor/statistics.py

```diff
@@ -0,0 +1,153 @@
+            self._name = '{} {}'.format(name, ATTR_COUNT)
+        self._sampling_size = sampling_size
+        self._unit_of_measurement = None
+        self.states = []
+        self.median = self.mean = self.variance = self.standard_deviation = 0
+        self.min = self.max = self.total = self.count = 0
+        self.update()
+
+        def calculate_sensor_state_listener(entity, old_state, new_state):
+            """Called when the sensor changes state."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+
+            try:
+                state = float(new_state.state)
+
+                if self._sampling_size is 0:
+                    self.states.append(state)
+                elif len(self.states) is self._sampling_size:
+                    del self.states[0]
+                    self.states.append(state)
+                else:
+                    self.states.append(state)
+            except ValueError:
+                self.count = self.count + 1
+
+            self.update_ha_state(True)
+
+        track_state_change(hass, entity_id, calculate_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self.mean if not self.is_binary else self.count
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit_of_measurement if not self.is_binary else None
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of the sensor."""
+        if not self.is_binary:
+            return {
+                ATTR_MEAN: self.mean,
+                ATTR_COUNT: self.count,
+                ATTR_MAX_VALUE: self.max,
+                ATTR_MEDIAN: self.median,
+                ATTR_MIN_VALUE: self.min,
+                ATTR_SAMPLING_SIZE: 'unlimited' if self._sampling_size is
+                                    0 else self._sampling_size,
+                ATTR_STANDARD_DEVIATION: self.standard_deviation,
+                ATTR_TOTAL: self.total,
+                ATTR_VARIANCE: self.variance,
+            }
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return ICON
+
+    def update(self):
+        """Get the latest data and updates the states."""
+        try:
+            self.mean = round(statistics.mean(self.states), 2)
+            self.median = round(statistics.median(self.states), 2)
+            self.standard_deviation = round(statistics.stdev(self.states), 2)
+            self.total = round(sum(self.states), 2)
+            self.variance = round(statistics.variance(self.states), 2)
+        except statistics.StatisticsError as err:
```

Comment:
In the start phase when there are no enough values available for the calculations. A lot of  functions from `statistics` requires at least two values.